presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java

```diff
@@ -273,19 +232,42 @@ private static Response toResponse(Query query, QueryResults queryResults)
     }
 
     @DELETE
-    @Path("{queryId}/{token}")
+    @Path("/v1/statement/executing/{queryId}/{token}")
     @Produces(MediaType.APPLICATION_JSON)
     public Response cancelQuery(
             @PathParam("queryId") QueryId queryId,
             @PathParam("token") long token,
             @QueryParam("slug") String slug)
     {
-        Query query = getQuery(queryId, slug);
-        if (query == null) {
-            return Response.status(Status.NOT_FOUND).build();
+        Query query = queries.get(queryId);
+        if (query != null) {
+            if (!query.isSlugValid(slug)) {
+                throw badRequest(NOT_FOUND, "Query not found");
+            }
+            query.cancel();
+            return Response.noContent().build();
+        }
+
+        // cancel the query execution directly instead of creating the statement client
+        try {
+            if (!queryManager.isQuerySlugValid(queryId, slug)) {
+                throw badRequest(NOT_FOUND, "Query not found");
+            }
+            queryManager.cancelQuery(queryId);
+            return Response.noContent().build();
+        }
+        catch (NoSuchElementException e) {
+            throw badRequest(NOT_FOUND, "Query not found");
         }
-        query.cancel();
-        return Response.noContent().build();
+    }
+
+    private static WebApplicationException badRequest(Status status, String message)
```

Comment:
What I'll do is, because every method that calls this passes in `NOT_FOUND` for the status, is I'll hardcode that `NOT_FOUND` and rename to `notFound`.